lib/adf_derive.py

```diff
@@ -0,0 +1,274 @@
+    # No time series creation
+    exit_msg = f"WARNING: {var} is not in the file {hist0} and can't be derived."
+    exit_msg += "\t     ** No time series will be generated. **"
+
+    # Initialiaze list for constituents
+    # NOTE: This is if the variable is NOT derivable but needs
+    #       an empty list as a check later
+    constit_list = []
+
+    try_cam_constits = True
+    # Try finding info from variable defaults yaml file
+    try:
+        vres = res[var]
+    except KeyError:
+        print(exit_msg)
+        self.debug_log(exit_msg)
+        return diag_var_list, constit_dict
+
+    # Check first if variable is potentially part of a CAM-CHEM run
+    if "derivable_from_cam_chem" in vres:
+        constit_list = vres["derivable_from_cam_chem"]
+
+        if constit_list:
+            if all(item in hist_file_ds.data_vars for item in constit_list):
+                # Set check to look for regular CAM constituents in variable defaults
+                try_cam_constits = False
+                msg = f"derive time series for {case_name}:"
+                msg += "\n\tLooks like this a CAM-CHEM run, "
+                msg += f"checking constituents for '{var}'"
+                self.debug_log(msg)
+        else:
+            self.debug_log(constit_errmsg)
+        # End if
+    # End if
+    
+    # If not CAM-CHEM, check regular CAM runs
+    if try_cam_constits:
+        if "derivable_from" in vres:
+            constit_list = vres["derivable_from"]
+        else:
+            # Missing variable or missing derivable_from argument
+            der_from_msg = f"derive time series for {case_name}:"
+            der_from_msg += f"\n Can't create time series for {var}.\n\tEither "
+            der_from_msg += "the variable is missing from CAM output or it is a "
+            der_from_msg += "derived quantity and is missing the 'derivable_from' "
+            der_from_msg += "config argument.\n\tPlease add variable to CAM run "
+            der_from_msg += "or set appropriate argument in variable "
+            der_from_msg += "defaults yaml file."
+            self.debug_log(der_from_msg)
+        # End if
+    # End if
+
+    # Log if this variable can be derived but is missing list of constituents
+    if isinstance(constit_list, list) and not constit_list:
+        self.debug_log(constit_errmsg)
+
+    # Check if any constituents were found
+    if constit_list:
+        # Add variable and constituent list to dictionary
+        constit_dict[var] = constit_list
+
+        # Add constituents to ADF diag variable list for time series generation
+        for constit in constit_list:
+            if constit not in diag_var_list:
+                diag_var_list.append(constit)
+    else:
+        print(exit_msg)
+        self.debug_log(exit_msg)
+    # End if
+
+    return diag_var_list, constit_dict
+
+########
+
+def derive_variable(self, case_name, var, res=None, ts_dir=None,
+                         constit_list=None, overwrite=None):
```

Comment:
I think all of these variables should be required except `overwrite`, as the code will fail as-is if those quantities aren't provided.  Also we should rename `self` to something like `adfobj`, at least until it becomes part of the offficial ADF class structure:

```suggestion
def derive_variable(adfobj, case_name, var, res, ts_dir,
                         constit_list, overwrite=False):
```